src/Tools/dotnet-monitor/CollectionRules/ActionOptionsDependencyAnalyzer.cs

```diff
@@ -0,0 +1,255 @@
+        private const string SubstitutionPrefix = "$(";
+        private const string SubstitutionSuffix = ")";
+        private const string Separator = ".";
+        private const string ActionReferencePrefix = SubstitutionPrefix + "Actions.";
+
+        private readonly CollectionRuleContext _ruleContext;
+
+        //Use action index instead of name, since it's possible for an unnamed action to have named dependencies.
+        private Dictionary<int, Dictionary<string, PropertyDependency>> _dependencies;
+
+        private sealed class PropertyDependency
+        {
+            public PropertyDependency(PropertyInfo property, string originalValue)
+            {
+                Property = property;
+                OriginalValue = originalValue;
+            }
+
+            public PropertyInfo Property { get; }
+
+            public string OriginalValue { get; }
+
+            public List<ActionDependency> ActionDependencies = new();
+        }
+
+        private sealed class ActionDependency
+        {
+            public ActionDependency(CollectionRuleActionOptions action, string resultName)
+            {
+                Action = action;
+                ResultName = resultName;
+            }
+
+            public CollectionRuleActionOptions Action { get; }
+
+            public string ResultName { get; }
+
+            public string GetActionResultToken() => string.Concat(ActionReferencePrefix,
+                Action.Name, Separator, ResultName, SubstitutionSuffix);
+        }
+
+        public ActionOptionsDependencyAnalyzer(CollectionRuleContext context)
+        {
+            _ruleContext = context ?? throw new ArgumentNullException(nameof(context));
+        }
+
+        public IList<CollectionRuleActionOptions> GetActionDependencies(int actionIndex)
+        {
+            EnsureDependencies();
+
+            if (_dependencies.TryGetValue(actionIndex, out Dictionary<string, PropertyDependency> properties))
+            {
+                return properties
+                    .SelectMany(p => p.Value.ActionDependencies)
+                    .DistinctBy(a => a.Action.Name, StringComparer.OrdinalIgnoreCase)
+                    .Select(a => a.Action).ToArray();
+            }
+            return Array.Empty<CollectionRuleActionOptions>();
+        }
+
+        public object SubstituteOptionValues(int actionIndex, object settings)
+        {
+            EnsureDependencies();
+            if (!_dependencies.TryGetValue(actionIndex, out Dictionary<string, PropertyDependency> properties))
+            {
+                return settings;
+            }
+
+            if (settings is ICloneable cloneable)
+            {
+                settings = cloneable.Clone();
+            }
+            else
+            {
+                _ruleContext.Logger.InvalidSettings(settings.GetType().FullName);
+                return settings;
+            }
+
+            foreach (KeyValuePair<string, PropertyDependency> property in properties)
```

Comment:
nit: you can use decomposition to make it more clear what each part is:

```suggestion
            foreach ((string propertyName, PropertyDependency dependency) in properties)
```